Data race?

I found this error in my main code , relevant log is : 

```
2015/09/07 12:37:01 [STOP - Hammer Time] Forcefully shutting down parent
2015/09/07 12:37:01 WaitGroup at 0 sync: negative WaitGroup counter
2015/09/07 12:37:01 22203 Serve() returning...
panic: sync: WaitGroup is reused before previous Wait has returned

goroutine 1 [running]:
sync.(*WaitGroup).Wait(0xc82020aee0)
        /usr/local/go/src/sync/waitgroup.go:128 +0x114
github.com/fvbock/endless.(*endlessServer).Serve(0xc82020ae60, 0x7f9fbe8b1500, 0xc82059f130)
        /home/azmoona/app_build/vendor/src/github.com/fvbock/endless/endless.go:166 +0x32c
github.com/fvbock/endless.(*endlessServer).ListenAndServe(0xc82020ae60, 0x0, 0x0)
        /home/azmoona/app_build/vendor/src/github.com/fvbock/endless/endless.go:197 +0x42a
github.com/fvbock/endless.ListenAndServe(0xc82022a900, 0xe, 0x7f9fbe8b3740, 0xc820242300, 0x0, 0x0)
        /home/azmoona/app_build/vendor/src/github.com/fvbock/endless/endless.go:136 +0x59
main.main()
        /home/azmoona/app_build/src/webserver/azmoona.go:58 +0x605
```

build the simple example with `-race` flag, and there is always a race. 

```
 ./simple 
2015/09/07 13:10:03 10751 localhost:4242
2015/09/07 13:10:15 10751 Received SIGHUP. forking.
2015/09/07 13:10:15 10751 Received SIGTERM.
==================
WARNING: DATA RACE
Read by goroutine 7:
  github.com/fvbock/endless.(*endlessServer).shutdown()
      /home/f0rud/gospace/src/github.com/fvbock/endless/endless.go:342 +0x57
  github.com/fvbock/endless.(*endlessServer).handleSignals()
      /home/f0rud/gospace/src/github.com/fvbock/endless/endless.go:316 +0x1272

Previous write by main goroutine:
  github.com/fvbock/endless.(*endlessServer).Serve()
      /home/f0rud/gospace/src/github.com/fvbock/endless/endless.go:163 +0x1f1
  github.com/fvbock/endless.(*endlessServer).ListenAndServe()
      /home/f0rud/gospace/src/github.com/fvbock/endless/endless.go:197 +0x526
  github.com/fvbock/endless.ListenAndServe()
      /home/f0rud/gospace/src/github.com/fvbock/endless/endless.go:136 +0x6a
  main.main()
      /home/f0rud/gospace/src/github.com/fvbock/endless/examples/simple.go:21 +0x220

Goroutine 7 (running) created at:
  github.com/fvbock/endless.(*endlessServer).ListenAndServe()
      /home/f0rud/gospace/src/github.com/fvbock/endless/endless.go:182 +0xbb
  github.com/fvbock/endless.ListenAndServe()
      /home/f0rud/gospace/src/github.com/fvbock/endless/endless.go:136 +0x6a
  main.main()
      /home/f0rud/gospace/src/github.com/fvbock/endless/examples/simple.go:21 +0x220
==================
2015/09/07 13:10:15 10751 127.0.0.1:4242 Listener closed.
2015/09/07 13:10:15 10751 Waiting for connections to finish...
2015/09/07 13:10:15 10751 Serve() returning...
2015/09/07 13:10:15 accept tcp 127.0.0.1:4242: use of closed network connection
2015/09/07 13:10:15 Server on 4242 stopped
2015/09/07 13:10:15 10930 localhost:4242
Found 1 data race(s)
```

--- a/endless.go
+++ b/endless.go
@@ -69,6 +69,7 @@ type endlessServer struct {
 	sigChan          chan os.Signal
 	isChild          bool
 	state            uint8
+	lock             *sync.RWMutex
 }
 
 /*
@@ -112,6 +113,7 @@ func NewServer(addr string, handler http.Handler) (srv *endlessServer) {
 			},
 		},
 		state: STATE_INIT,
+		lock:  &sync.RWMutex{},
 	}
 
 	srv.Server.Addr = addr
@@ -148,6 +150,20 @@ func ListenAndServeTLS(addr string, certFile string, keyFile string, handler htt
 	return server.ListenAndServeTLS(certFile, keyFile)
 }
 
+func (srv *endlessServer) getState() uint8 {
+	srv.lock.RLock()
+	defer srv.lock.RUnlock()
+
+	return srv.state
+}
+
+func (srv *endlessServer) setState(st uint8) {
+	srv.lock.Lock()
+	defer srv.lock.Unlock()
+
+	srv.state = st
+}
+
 /*
 Serve accepts incoming HTTP connections on the listener l, creating a new
 service goroutine for each. The service goroutines read requests and then call
@@ -160,11 +176,11 @@ down the server.
 */
 func (srv *endlessServer) Serve() (err error) {
 	defer log.Println(syscall.Getpid(), "Serve() returning...")
-	srv.state = STATE_RUNNING
+	srv.setState(STATE_RUNNING)
 	err = srv.Server.Serve(srv.EndlessListener)
 	log.Println(syscall.Getpid(), "Waiting for connections to finish...")
 	srv.wg.Wait()
-	srv.state = STATE_TERMINATE
+	srv.setState(STATE_TERMINATE)
 	return
 }
 
@@ -339,11 +355,11 @@ starts a goroutine that will hammer (stop all running requests) the server
 after DefaultHammerTime.
 */
 func (srv *endlessServer) shutdown() {
-	if srv.state != STATE_RUNNING {
+	if srv.getState() != STATE_RUNNING {
 		return
 	}
 
-	srv.state = STATE_SHUTTING_DOWN
+	srv.setState(STATE_SHUTTING_DOWN)
 	if DefaultHammerTime >= 0 {
 		go srv.hammerTime(DefaultHammerTime)
 	}
@@ -375,13 +391,13 @@ func (srv *endlessServer) hammerTime(d time.Duration) {
 			log.Println("WaitGroup at 0", r)
 		}
 	}()
-	if srv.state != STATE_SHUTTING_DOWN {
+	if srv.getState() != STATE_SHUTTING_DOWN {
 		return
 	}
 	time.Sleep(d)
 	log.Println("[STOP - Hammer Time] Forcefully shutting down parent")
 	for {
-		if srv.state == STATE_TERMINATE {
+		if srv.getState() == STATE_TERMINATE {
 			break
 		}
 		srv.wg.Done()
@@ -503,9 +519,9 @@ type endlessConn struct {
 }
 
 func (w endlessConn) Close() error {
-    err := w.Conn.Close()
-    if err == nil {
-        w.server.wg.Done()
-    }
-    return err
+	err := w.Conn.Close()
+	if err == nil {
+		w.server.wg.Done()
+	}
+	return err
 }

